File: table.js
import { shuffle, sample_indices, remove_by_idxs } from "./helpers.js";

export const SUITS = ["♠", "♣", "♦", "♥"];
export const RANKS = [
  "2",
  "3",
  "4",
  "5",
  "6",
  "7",
  "8",
  "9",
  "X",
  "J",
  "Q",
  "K",
  "A",
];

export class Table {
  constructor(player_ids) {
    this.deck = [];
    for (const suit of SUITS) {
      for (const rank of RANKS) {
        this.deck.push(`${suit}${rank}`);
      }
    }

    this.card_values = new Map();
    this.deck
      .filter((card) => card[0] === "♥")
      .forEach((card) => this.card_values.set(card, 1));
    this.deck
      .filter((card) => card[0] !== "♥")
      .forEach((card) => this.card_values.set(card, 0));
    this.card_values.set("♠Q", 13);

    this.hands = new Map();
    player_ids.forEach((player_id) => this.hands.set(player_id, []));

    this.cards_taken = new Map();
    player_ids.forEach((player_id) => this.cards_taken.set(player_id, []));
  }

  deal_cards(n, player_id) {
    const idxs = sample_indices(this.deck.length, n);
    for (const idx of idxs) {
      var card = this.deck[idx];
      this.hands.get(player_id).push(card);
    }
    this.deck = remove_by_idxs(this.deck, idxs);
  }

  update_hands(updates) {
    for (const [player_id, update] of updates.entries()) {
      this.hands.set(
        player_id,
        this.hands.get(player_id).concat(update.get("added"))
      );

      for (const card of update.get("removed")) {
        var idx = this.hands.get(player_id).indexOf(card);
        this.hands.get(player_id).splice(idx, 1);
      }
    }
  }
}

// const table = new Table(["Mary", "John"]);
// table.deal_cards(10, "Mary");
// table.deal_cards(10, "John");

// const updates = new Map([
//   [
//     "Mary",
//     new Map([
//       ["added", ["♥9", "♦6", "♥6"]],
//       ["removed", table.hands.get("Mary")],
//     ]),
//   ],
// ]);

// table.update_hands(updates);

// console.log(table.hands);

File: python_proto/table.py
from collections import namedtuple
import random


class Card(namedtuple("Card", ("suit", "rank"))):
    def __repr__(self) -> str:
        return f"{self.suit}{self.rank}"


class Play(namedtuple("Play", ("played_by", "card"))):
    def __repr__(self) -> str:
        return f"{self.played_by}/{self.card}"


class Table:
    def __init__(self, player_ids: list[str]) -> None:
        self.suits = {"♠", "♣", "♦", "♥"}
        self.ranks = [str(i) for i in list(range(2, 11)) + ["J", "Q", "K", "A"]]
        
        self.deck = {Card(suit, rank) for suit in self.suits for rank in self.ranks}
        self.card_values = {card: 1 if card.suit == "♥" else 0 for card in self.deck}
        self.card_values[Card("♠", "Q")] = 13

        self.hands: dict[str, set[Card]] = {player_id: set() for player_id in player_ids}
        self.cards_taken: dict[str, set[Card]] = {player_id: set() for player_id in player_ids}

    def deal_cards(self, n: int, player_id: str) -> None:
        dealt_cards = random.sample(sorted(self.deck), n)

        for card in dealt_cards:
            self.deck.remove(card)
            self.hands[player_id].add(card)

    def update_hands(self, updates: dict[str, dict[str, set[Card]]]) -> None:
        # {"Tom": {"added": [], "removed": []}}
        for player_id, update in updates.items():
            added = update.get("added", set())
            removed = update.get("removed", set())
            assert not self.hands[player_id].intersection(added)
            self.hands[player_id] |= added
            assert not removed - self.hands[player_id]
            self.hands[player_id] -= removed
